File: 2º Periodo/AEDS-II/Códigos/JAVA/ListaLinear.java
import java.util.Scanner;


class Ponteiro {
	int n;
}

public class Main
{

	public static void inverter(int[] lista, Ponteiro quantidade, int tamanho) {
	    
	    
	    if( quantidade.n == 0){
	        System.out.println("Lista vazia!!");
	        return;
	    }
	    

		int[] inverso = new int[quantidade.n]; // cria uma nova lista para armazenar os elementos inversos
		int j = quantidade.n - 1;

		for(int i = 0; i < quantidade.n; i++) {


			inverso[i] = lista[j];
			j--;
		}

	        mostrar(inverso, quantidade); // chama a funão para mostrar a lista inversa

	}



	public static void pesquisar(int numero, int[] lista, Ponteiro quantidade) {

		boolean achou = false;
		for (int i = 0; i < quantidade.n; i++) { // percorre toda a lista 

			if(lista[i] == numero) {
				System.out.println("Numero encontrado na pos: " + i);
				i = quantidade.n;
				achou = true;
			}


		}
		if(achou != true) {
			System.out.println("Numero não encontrado!");

		}

	}



	public static void mostrar(int[] lista, Ponteiro quantidade) {

		System.out.print("[");
		for (int i = 0; i < quantidade.n; i++) { // for de 0 a quantidade.n (onde tem elementos)
			System.out.print(lista[i]);
			System.out.print(" ");
		}
		System.out.println("]");
	}

	public static void remover(int pos, int[] lista, Ponteiro quantidade, int tamanho) {


			if( pos < 0 || pos > quantidade.n) { // verifica se  tem espaço, se pos não e negativo, se pos não é maior que a lista
			System.out.println("Lista cheia ou pos invalida!");
			return;
		}

		quantidade.n--;     // diminui o quantidade.n, para tratar o elemento em questão como lixo

		for(int i = pos; i < quantidade.n; i++) {  // move os elementos do final ate a posição excluida


			lista[i] = lista[i+1];
		}




	}



	public static int removerFim(int[] lista, Ponteiro quantidade, int tamanho) {


			if(quantidade.n  == 0  ){
			System.out.println("Lista cheia ou pos inválida!");
			return 0;
		}

		int removido = lista[quantidade.n];
		quantidade.n--;


		return removido;


	}



	public static int removerInicio(int[] lista, Ponteiro quantidade, int tamanho) {


			if(quantidade.n  == 0  ) { // verifica se a lista não esta vazia
			System.out.println("Lista cheia ou pos inválida!");
			return 0 ;
		}

		int removido = lista[0];
		quantidade.n--;


		for(int i = 0; i < quantidade.n; i++) {


			lista[i] = lista[i+ 1];
		}

		return removido;


	}


	public static void inserir(int pos, int numero, int[] lista, Ponteiro quantidade, int tamanho) {

		try {
			if(quantidade.n  >= tamanho || pos < 0 || pos > quantidade.n) { // verifica se  tem espaço, se pos não e negativo, se pos não C) maior que a lista

			}
		} catch(ArrayIndexOutOfBoundsException e) {
			System.out.println("Lista cheia ou pos inválida!");
		}


		for(int i = quantidade.n; i > pos; i--) {


			lista[i] = lista[i-1];
		}

		lista[pos] = numero;
		quantidade.n++;


	}



	public static void inserirFim( int numero, int[] lista, Ponteiro quantidade, int tamanho) {


			if(quantidade.n  >= tamanho) {
			System.out.println("Lista cheia!");
			return;
		}

		lista[quantidade.n] = numero;
		quantidade.n++;
	}
	

	public static void inserirInicio(int numero, int[] lista, Ponteiro quantidade, int tamanho) {


			if(quantidade.n >= tamanho) {
			System.out.println("Lista cheia!");
			return;
		}

		for(int i = quantidade.n; i > 0; i--) {

			lista[i] = lista[i - 1];
		}

		lista[0] = numero;
		quantidade.n++;
	}


	public static void main(String[] args) {
		Scanner Scanner = new Scanner (System.in);
		boolean isfim = false;
		Ponteiro ponteiro = new Ponteiro();
		ponteiro.n = 0;
		int[] lista = new int[100];

		do {

			System.out.println("--------------------------------------------------------------------------");
			System.out.println("\nbem vindo aos exemplos de TAD, aqui temos algumas opções para treinar \ntemos as opções:\n a - inserir no inicio\n b - inserir no fim\n c - inserir em alguma posição\n d - remover no inicio\n e - remover no fim\n f - remover\n g - mostrar\n h - pesquisar\n i - Inverter elementos\n FIM - para sair do programa\n ");
			System.out.println("--------------------------------------------------------------------------");

			System.out.print("digite a escolha que deseja:");
			String escolha = Scanner.nextLine();


			int numero = 0;
			int pos = 0;

			if(escolha.equalsIgnoreCase("FIM")) {
				isfim = true;
				break;
			}


			switch (escolha) {


			case "a":
				System.out.print("Digite o numero desejado: ");
				numero = Scanner.nextInt();
				Scanner.nextLine();
				inserirInicio(numero, lista, ponteiro, lista.length);
				mostrar(lista, ponteiro);

				break;

			case "b":
				System.out.print("Digite o numero desejado: ");
				numero = Scanner.nextInt();
				Scanner.nextLine();
				inserirFim(numero, lista, ponteiro, lista.length);
				mostrar(lista, ponteiro);

				break;

			case "c":
				System.out.print("Digite o numero desejado: ");
				numero = Scanner.nextInt();
				Scanner.nextLine();
				System.out.print("Digite a posição desejada: ");
				pos = Scanner.nextInt();
				Scanner.nextLine();
				inserir(pos, numero, lista, ponteiro, lista.length);
				mostrar(lista, ponteiro);
				break;

			case "d":

				int removido = removerInicio(lista, ponteiro, lista.length);
				System.out.println("elemento removido: " + removido);
				mostrar(lista, ponteiro);
				break;


			case "e":
				removido = removerFim(lista, ponteiro, lista.length);
				mostrar(lista, ponteiro);
				break;


			case "f":

				System.out.print("Digite a posição desejada: ");
				pos = Scanner.nextInt();
				Scanner.nextLine();
				remover(pos, lista, ponteiro, lista.length);
				mostrar(lista, ponteiro);
				break;


			case "g":
				mostrar(lista, ponteiro);
				break;

			case "h":
				System.out.print("Digite a o numero desejado: ");
				numero = Scanner.nextInt();
				Scanner.nextLine();
				pesquisar(numero, lista, ponteiro);
				break;

			case "i":
				inverter(lista, ponteiro, lista.length);
				break;

			default:
				System.out.print("Invalido!");

			}


		} while(isfim != true);
		Scanner.close();
	}
}

File: 2º Periodo/AEDS-II/Códigos/C/ListaLinear.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

void remover(int *n, int *vetor, int tamanho, int pos) {
    if (pos < 0 || pos >= *n) { // Corrigida a condição de validação
        printf("Erro: posição inválida!\n");
        return;
    }

    int removido = vetor[pos]; // Armazena o elemento removido
    (*n)--; // Corrige a decretação do tamanho da lista

    for (int i = pos; i < *n; i++) { // Desloca os elementos
        vetor[i] = vetor[i + 1];
    }

    printf("Removido: %d\n", removido); // Exibe o elemento removido
}



void removerFim(int *n) {
    if (*n == 0) { // Verifica se a lista está vazia
        printf("Erro: lista vazia\n");
        return;
    }
    --*n;
}

void removerInicio(int *n, int *vetor, int tamanho) {
    if (*n == 0) { // Verifica se a lista está vazia
        printf("Erro: lista vazia\n");
        return;
    }

    int removido = vetor[0];  // Armazena o elemento removido para exibição
    (*n)--;

    for (int i = 0; i < *n; i++) { // Desloca os elementos
        vetor[i] = vetor[i + 1];
    }

}


void inserirInicio(int x, int *n, int *vetor, int tamanho){
    
    
    if(*n >= tamanho){ // verifica se a lista não esta cheia
        printf("erro, sem espaço\n");
        return; 
    }
    
    for(int i= *n; i > 0; i--){ // desloca o elemento para deixar o inicio livre
        
        vetor[i] = vetor[i - 1];
        
    }
    
    vetor[0] = x; // adiciona o numero desejado na 1º pos
    (*n)++; //incrementa o n para a prox pos desocupada
    
}

void inserirFim(int x, int *n, int *vetor, int tamanho){
    
    
    if(*n >= tamanho){ // verifica se a lista não esta cheia
        printf("erro, sem espaço\n");
        return; 
    }
    
   
    vetor[*n] = x; // adiciona o numero desejado na 1º pos
    (*n)++; //incrementa o n para a prox pos desocupada
    
}

void inserir(int x, int *n, int *vetor, int tamanho, int pos){
    
    
    if(*n >= tamanho || pos < 0  || pos > *n){ // verifica se a lista não esta cheia, se a pos n é negativa e se a pos não e maior que a lista
        printf("erro, sem espaço ou posição invalida!\n");
        return; 
    }
    
      for(int i= *n; i > pos; i--){ // desloca o elemento para deixar o inicio livre
        
        vetor[i] = vetor[i - 1];
        
    }
    
    vetor[pos] = x; // adiciona o numero desejado na 1º pos
    (*n)++; //incrementa o n para a prox pos desocupada
    
}


int main()
{
    int vet[10]; // lista linear com espaço para 11 elementos (0,1,2,3,4,5,6,7,8,9,10)
    int n = 0; // contador, para saber onde tem espaço livre na lista
    while(1){
        
        printf("\nbem vindo aos exemplos de TAD, aqui temos algumas opções para treinar \ntemos as opções:\n a - inserir no inicio\n b - inserir no fim\n c - inserir em alguma posição\n d - remover no inicio\n e - remover no fim\n g - remover\n FIM - para sair do programa\n ");
        char palavra[5];
        fgets(palavra,sizeof(palavra), stdin);
        palavra[strcspn(palavra, "\n")] = '\0';
        int numero; // ultilizado para identificar o valor a ser inserido
		if (strcmp(palavra, "FIM") == 0) {  // caso a palavra for FIM, o programa para
			break;
		}
        
        switch (palavra[0]) {
        case 'a':
        
             numero = 0;
            scanf("%d", &numero);
            getchar();
            inserirInicio(numero, &n, vet, 10); /* n sendo passado como ponteiro pois se passar como valor
        a função cria uma copia para n que não é alterada aqui na Main*/
            break;
            
        case 'b':
             numero = 0;
            scanf("%d", &numero);
            getchar();
            inserirFim(numero, &n, vet, 10); /* n sendo passado como ponteiro pois se passar como valor
        a função cria uma copia para n que não é alterada aqui na Main*/
            break;
            
        case 'c':
            numero = 0;
            scanf("%d", &numero);
            int pos = scanf("%d" , &pos);
            getchar();
            inserir(numero, &n, vet, 10,  pos);
            break;
            
        case 'd':
            removerInicio(&n, vet, 10);
            break;
            
        case 'e':
             removerFim(&n);
            break;
        case 'f':
            printf("Sexta-feira\n");
            break;
        case 'g':
            pos = scanf("%d" , &pos);
            getchar();
            remover(&n, vet, 10,  pos);
            break;
        default:
            printf("inválido!\n");
    }
        
        

        for (int i = 0; i < n; i++) {
         printf("%d ", vet[i]);
        }
    
    }
    return 0;
}
